src/adr/0036-against-use-of-conditional-statements-in-guice-modules.md

```diff
@@ -0,0 +1,118 @@
+[Cassandra BlobStore cache](0025-cassandra-blob-store-cache.md) proposed to leverage Cassandra as a performance 
+(latency) enhancer for ObjectStorage technologies.
+
+Yet again it had been decided to use conditional statements in modules in order to lower the number of products.
+
+However, some components requires expensive resource initialization. These operations are performed via a separate module
+that needs to be installed based on the configuration. For instance 
+[Cassandra BlobStore cache](0025-cassandra-blob-store-cache.md) requires usage of an additional cache keyspace that 
+represents a cost and an inconvenience we don't want to pay if we don't rely on that cache. Not having the cache module 
+thus enables quickly auditing that the caching cassandra session is not initialized. See 
+[this comment](https://github.com/linagora/james-project/pull/3261#pullrequestreview-389804841) as well as 
+[this comment](https://github.com/linagora/james-project/pull/3261#issuecomment-613911695).
+
+### Audit
+
+The following modules perform conditional statements upon injection time:
+
+ - BlobExportMechanismModule : Choice of the export mechanism
+ - ObjectStorageDependenciesModule::selectBlobStoreBuilder: Choice between S3 and Swift ObjectStorage technologies
+ - TikaMailboxModule::provideTextExtractor: Choice of text extraction technology
+ - BlobStoreChoosingModule::provideBlobStore: Choice of BlobStore technology: Cassandra, ObjectStorage or Hybrid
+ - [Cached blob store](https://github.com/linagora/james-project/pull/3319) represents a similar problem: should the 
+ blobStore be wrapped by a caching layer?
+
+## Decision
+
+We should no longer rely on conditional statements in Guice module.
+
+Guice modules combination choice should be decided before starting the dependency injection stage.
+
+An INFO log with the list of modules used to create its Guice injector. This enables easy diagnose of the running 
+components via the selected module list. It exposes tested, safe choices to the user while limiting the Guice products 
+count.
+
+The following conditional statements in guice modules needs to be removed :
+
+ - [Cached blob store pull request](https://github.com/linagora/james-project/pull/3319) addresses 
+ ObjectStorageDependenciesModule::selectBlobStoreBuilder and Cassandra Blob Store Cache conditional statement.
+ - [S3 native blobStore implementation](https://github.com/linagora/james-project/pull/3099) along side with S3 endpoints
+ support as part of Swift removes the need to select the Object Storage implementation.
+ - Follow up work needs to be plan concerning `BlobExportMechanismModule` and `TikaMailboxModule::provideTextExtractor`.
+
+## Consequences
+
+Component combination count keeps unchanged for Guice products, but the run combination is explicit. QA needs are 
+unchanged.
+
+Integration testing can not offer conditional, configuration based module composition. This is because:
```

Comment:
I would really like we don't  do that to prevent diverging from what happens in production.
I'd like that:
* we define POJOs for configurations involving module choosing with a nice programmatic interface to build them
* we create a method that takes a module combination + POJOs to start the server and rely on that for both tests and production
* we write parsers for config file to POJOs and call the method to start the server once the POJOs are built

WDYT?